presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyRowExpressions.java

```diff
@@ -0,0 +1,92 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.LogicalRowExpressions;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.RowExpressionRewriter;
+import com.facebook.presto.spi.relation.RowExpressionTreeRewriter;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.type.BooleanType;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.relational.FunctionResolution;
+import com.facebook.presto.sql.relational.RowExpressionDeterminismEvaluator;
+import com.facebook.presto.sql.relational.RowExpressionOptimizer;
+
+import static com.facebook.presto.spi.relation.ExpressionOptimizer.Level.SERIALIZABLE;
+import static java.util.Objects.requireNonNull;
+
+public class SimplifyRowExpressions
+        extends RowExpressionRewriteRuleSet
+{
+    public SimplifyRowExpressions(Metadata metadata)
+    {
+        super(new Rewriter(metadata));
+    }
+
+    private static class Rewriter
+            implements PlanRowExpressionRewriter
+    {
+        private final Metadata metadata;
+        private final RowExpressionOptimizer optimizer;
+
+        public Rewriter(Metadata metadata)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.optimizer = new RowExpressionOptimizer(metadata);
+        }
+
+        @Override
+        public RowExpression rewrite(RowExpression expression, Rule.Context context)
+        {
+            RowExpression optimized = optimizer.optimize(expression, SERIALIZABLE, context.getSession().toConnectorSession());
+            if (optimized instanceof ConstantExpression || !BooleanType.BOOLEAN.equals(optimized.getType())) {
+                return optimized;
+            }
+            return RowExpressionTreeRewriter.rewriteWith(new LogicalExpressionRewriter(metadata), optimized);
+        }
+    }
+
+    private static class LogicalExpressionRewriter
+            extends RowExpressionRewriter<Void>
+    {
+        LogicalRowExpressions logicalRowExpressions;
+
+        public LogicalExpressionRewriter(Metadata metadata)
+        {
+            this.logicalRowExpressions = new LogicalRowExpressions(new RowExpressionDeterminismEvaluator(metadata.getFunctionManager()), new FunctionResolution(metadata.getFunctionManager()));
+        }
+
+        @Override
+        public RowExpression rewriteCall(CallExpression node, Void context, RowExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (BooleanType.BOOLEAN.equals(node.getType())) {
+                return logicalRowExpressions.pushNegationToLeaves(node);
```

Comment:
`pushDownNegations` is more sophisticated than `pushNegationToLeaves`. Could you check the consistency between these two and add tests?